### PR TITLE
kademlia persistence simulation

### LINKED ISSUE #262: Test for persisted state
There should be a test in discovery to see if the persisted peer set (1) is found and loaded (2) bootstraps healthy kademlia without any connections (3) benchmark the reduction of time reaching health

### PR DESCRIPTION
**Pull request by acud**
- fixed a bug in state store that did not close the file handle on shutdown
- added test for kademlia state storage across sessions

fixes #262

### CODE REVIEW

**Comment by acud** [comment]
the discovery test flakes at a rate of 3.5% locally. most probably due to a deadlock from mutual RPC dial-ins on the p2p package.

File: swarm/network/hive.go [hunk 1]
```diff
@@ -123,7 +125,14 @@ func (h *Hive) Stop() error {
 	log.Info(fmt.Sprintf("%08x hive stopping, saving peers", h.BaseAddr()[:4]))
 	h.ticker.Stop()
 	if h.Store != nil {
-		return h.savePeers()
+		if err := h.savePeers(); err != nil {
+			log.Error(fmt.Sprintf("could not save peers to persistence store: %v", err))
```
Comment (zelig) [inline swarm/network/hive.go]: remove log line.
its ungo to log error you return

File: swarm/network/hive.go [hunk 2]
```diff
@@ -141,6 +150,9 @@ func (h *Hive) Stop() error {
 func (h *Hive) connect() {
 	time.Sleep(time.Duration(rand.Intn(1000)) * time.Millisecond)
 	for range h.ticker.C {
+		time.Sleep(100 * time.Millisecond)
```
Comment (zelig) [inline swarm/network/hive.go]: remove sleeps

File: swarm/network/kademlia.go [hunk 3]
```diff
@@ -73,8 +73,8 @@ func NewKadParams() *KadParams {
 		MinProxBinSize: 2,
 		MinBinSize:     2,
 		MaxBinSize:     4,
-		RetryInterval:  4200000000, // 4.2 sec
-		MaxRetries:     42,
+		RetryInterval:  420000000, // 4.2 sec
```
Comment (zelig) [inline swarm/network/kademlia.go]: we should leave orig defaults

File: swarm/network/simulations/discovery/discovery_test.go [hunk 4]
```diff
@@ -20,23 +22,59 @@ import (
...
+	return store, nil
+}
+
 var (
-	nodeCount    = flag.Int("nodes", 16, "number of nodes to create (default 10)")
+	nodeCount    = flag.Int("nodes", 10, "number of nodes to create (default 10)")
 	initCount    = flag.Int("conns", 1, "number of originally connected peers	 (default 1)")
 	snapshotFile = flag.String("snapshot", "", "create snapshot")
-	loglevel     = flag.Int("loglevel", 3, "verbosity of logs")
+	loglevel     = flag.Int("loglevel", 6, "verbosity of logs")
```
Comment (zelig) [inline swarm/network/simulations/discovery/discovery_test.go]: leave 3 as default

File: swarm/network/simulations/discovery/discovery_test.go [hunk 5]
```diff
@@ -144,6 +190,29 @@ func testDiscoverySimulation(t *testing.T, nodes, conns int, adapter adapters.No
...
+	result, err := discoveryPersistenceSimulation(nodes, conns, adapter)
+
+	if err != nil {
+		t.Fatalf("Setting up simulation failed: %v", err)
+	}
+	if result.Error != nil {
+		t.Fatalf("Simulation failed: %s", result.Error)
+	}
+	t.Logf("Simulation with %d nodes passed in %s", nodes, result.FinishedAt.Sub(result.StartedAt))
+	cleanDbStores()
```
Comment (zelig) [inline swarm/network/simulations/discovery/discovery_test.go]: defer it after `discoveryPersistenceSimulation`

File: swarm/network/simulations/discovery/discovery_test.go [hunk 6]
```diff
@@ -266,6 +335,182 @@ func discoverySimulation(nodes, conns int, adapter adapters.NodeAdapter) (*simul
...
+
+		addrs = append(addrs, a)
+	}
+
+	// run a simulation which connects the 10 nodes in a ring and waits
+	// for full peer discovery
+	ppmap := network.NewPeerPot(testMinProxBinSize, ids, addrs)
+
+	var restartTime time.Time
+	var mutex = &sync.Mutex{}
```
Comment (zelig) [inline swarm/network/simulations/discovery/discovery_test.go]: need for lock?

File: swarm/network/simulations/discovery/discovery_test.go [hunk 7]
```diff
@@ -266,6 +335,182 @@ func discoverySimulation(nodes, conns int, adapter adapters.NodeAdapter) (*simul
...
+		Expect: &simulations.Expectation{
+			Nodes: ids,
+			Check: check,
+		},
+	})
+	if result.Error != nil {
+		return result, nil
+	}
+
+	if *snapshotFile != "" {
```
Comment (zelig) [inline swarm/network/simulations/discovery/discovery_test.go]: needed?

File: swarm/network/simulations/discovery/discovery_test.go [hunk 8]
```diff
@@ -266,6 +335,182 @@ func discoverySimulation(nodes, conns int, adapter adapters.NodeAdapter) (*simul
...
+			return false, fmt.Errorf("error getting node health: %s", err)
+		}
+		log.Info(fmt.Sprintf("node %4s healthy: got nearest neighbours: %v, know nearest neighbours: %v, saturated: %v", id, healthy.GotNN, healthy.KnowNN, healthy.Full))
+
+		return healthy.KnowNN && healthy.GotNN && healthy.Full, nil
+	}
+
+	// 64 nodes ~ 1min
+	// 128 nodes ~
+	timeout := 30 * time.Second
```
Comment (zelig) [inline swarm/network/simulations/discovery/discovery_test.go]: not too shorrt?

File: swarm/network/simulations/discovery/discovery_test.go [hunk 9]
```diff
@@ -313,27 +558,34 @@ func newService(ctx *adapters.ServiceContext) (node.Service, error) {
...
 	if ctx.Config.Reachable != nil {
 		kp.Reachable = func(o network.OverlayAddr) bool {
 			return ctx.Config.Reachable(o.(*network.BzzAddr).ID())
 		}
 	}
 	kad := network.NewKademlia(addr.Over(), kp)
-
 	hp := network.NewHiveParams()
-	hp.KeepAliveInterval = 200 * time.Millisecond
+	hp.KeepAliveInterval = time.Duration(200+rand.Intn(400)) * time.Millisecond
```
Comment (zelig) [inline swarm/network/simulations/discovery/discovery_test.go]: eliminate this indeterminacy

**Comment by zelig** [comment]
rebase on #379

File: swarm/network/simulations/discovery/discovery_test.go (hunk 6 above)
Comment (acud) [reply swarm/network/simulations/discovery/discovery_test.go]: the ticker loops at 200ms. 
the first block before `if isHealthy` would probably complete within a 200ms timeframe.
but the `isHealthy` block wouldn't, then the interrupt will be triggered again, and if there's no lock involved - the kademila first block will start executing again, then the second again as a result.
i'd rather leave the lock there

File: swarm/network/hive.go [hunk 10]
```diff
@@ -123,7 +124,13 @@ func (h *Hive) Stop() error {
 	log.Info(fmt.Sprintf("%08x hive stopping, saving peers", h.BaseAddr()[:4]))
 	h.ticker.Stop()
 	if h.Store != nil {
-		return h.savePeers()
+		if err := h.savePeers(); err != nil {
+			return fmt.Errorf("could not save peers to persistence store: %v", err)
+		}
+		if err := h.Store.Close(); err != nil {
+			log.Error(fmt.Sprintf("could not close file handle to persistence store: %v", err))
```
Comment (zelig) [inline swarm/network/hive.go]: i thought i commented on this already, do not log errors you return

File: swarm/network/hive.go [hunk 11]
```diff
@@ -139,8 +146,10 @@ func (h *Hive) Stop() error {
 // at each iteration, ask the overlay driver to suggest the most preferred peer to connect to
 // as well as advertises saturation depth if needed
 func (h *Hive) connect() {
-	time.Sleep(time.Duration(rand.Intn(1000)) * time.Millisecond)
+	//time.Sleep(time.Duration(rand.Intn(1000)) * time.Millisecond)
```
Comment (zelig) [inline swarm/network/hive.go]: remove commented code

File: swarm/network/simulations/discovery/discovery_test.go [hunk 12]
```diff
@@ -9,31 +9,70 @@ import (
...
 	"github.com/ethereum/go-ethereum/p2p/simulations/adapters"
 	"github.com/ethereum/go-ethereum/swarm/network"
+	"github.com/ethereum/go-ethereum/swarm/state"
 	colorable "github.com/mattn/go-colorable"
 )
 
 // serviceName is used with the exec adapter so the exec'd binary knows which
 // service to execute
 const serviceName = "discovery"
+const serviceNamePersistence = "noDiscovery_persistence"
```
Comment (zelig) [inline swarm/network/simulations/discovery/discovery_test.go]: `no_disc...`

File: swarm/network/simulations/discovery/discovery_test.go [hunk 13]
```diff
@@ -313,27 +547,34 @@ func newService(ctx *adapters.ServiceContext) (node.Service, error) {
...
 	if ctx.Config.Reachable != nil {
 		kp.Reachable = func(o network.OverlayAddr) bool {
 			return ctx.Config.Reachable(o.(*network.BzzAddr).ID())
 		}
 	}
 	kad := network.NewKademlia(addr.Over(), kp)
-
 	hp := network.NewHiveParams()
-	hp.KeepAliveInterval = 200 * time.Millisecond
+	//hp.KeepAliveInterval = time.Duration(200+rand.Intn(400)) * time.Millisecond
```
Comment (zelig) [inline swarm/network/simulations/discovery/discovery_test.go]: remove commented code

File: swarm/network/hive.go (hunk 10 above)
Comment (acud) [reply swarm/network/hive.go]: came back after rebasing

**Review by zelig — APPROVED** [review]

**Comment by holisticode** [comment]
Please consider that if someone requests my review, and then it gets closed/merged 7 hours later,  I might not have had the time to look at it, due to time zone difference

**Comment by acud** [comment]
@holisticode as I mentioned yesterday at the standup this is my mistake as I didn’t enforce the two reviews myself. Won’t happen again. Would still appreciate your review and will amend the changes in a refactoring PR. Sorry for the overhead.

**Pull request merged**